servicetalk-transport-netty-internal/src/test/java/io/servicetalk/transport/netty/internal/NettyChannelPublisherTest.java

```diff
@@ -111,6 +113,27 @@ public void tearDown() throws Exception {
         }
     }
 
+    @Test
+    public void testCancelThenResubscribeDeliversErrorAndNotQueuedData() throws InterruptedException {
+        TestCollectingPublisherSubscriber<Integer> subscriber1 = new TestCollectingPublisherSubscriber<>();
+        TestCollectingPublisherSubscriber<Integer> subscriber2 = new TestCollectingPublisherSubscriber<>();
+        toSource(publisher).subscribe(subscriber1);
+        Subscription subscription1 = subscriber1.awaitSubscription();
+        subscription1.request(1);
+
+        assertFalse(channel.writeInbound(1));
+        Integer next = subscriber1.takeOnNext();
+        assertThat(next, is(1));
+        assertFalse(channel.writeInbound(2)); // this write should be queued, because there isn't any requestN demand.
+
+        subscription1.cancel(); // cancel of active subscription should clear the queue and fail future Subscribers.
+
+        toSource(publisher).subscribe(subscriber2);
```

Comment:
Can you also add a test that does a `channelRead()` between `cancel()` and re-subscribe? That will re-initialize the `pending` queue.

Reply:
The `cancel` operation does a `shutdownInput` so no more data should be delivered by Netty via `channelRead`. `EmbeddedChannel` throws `ClosedChannelException` if I try to deliver more data, and I will add a test for this.

Reply:
The case I was trying to get to was when data is enqueued in the channel when `cancel()` happens in which case Netty should deliver data first and then the shutdown. We do not have to go out of our way to verify the scenario but looks like having a handler that buffers data could do the trick?

Reply:
good call ... I was trying to get away with avoiding additional conditional in `channelRead` but it is helpful to avoid this case.